paperqa/types.py

```diff
@@ -21,75 +11,204 @@
     select_paper_prompt,
     summary_prompt,
 )
-from .utils import extract_doi, iter_citations
 
+# Just for clarity
 DocKey = Any
-CBManager = Union[AsyncCallbackManagerForChainRun, CallbackManagerForChainRun]
-CallbackFactory = Callable[[str], Union[None, List[BaseCallbackHandler]]]
+
+CallbackFactory = Callable[[str], list[Callable[[str], None]] | None]
+
+
+class Embeddable(BaseModel):
+    embedding: list[float] | None = Field(default=None, repr=False)
 
 
-class Doc(BaseModel):
+class Doc(Embeddable):
     docname: str
     citation: str
     dockey: DocKey
 
 
-class Text(BaseModel):
+class Text(Embeddable):
     text: str
     name: str
     doc: Doc
-    embeddings: Optional[List[float]] = None
+
+
+def cosine_similarity(a, b):
+    dot_product = np.dot(a, b.T)
+    norm_product = np.linalg.norm(a, axis=1) * np.linalg.norm(b, axis=1)
+    return dot_product / norm_product
+
+
+class VectorStore(BaseModel, ABC):
+    """Interface for vector store - very similar to LangChain's VectorStore to be compatible"""
+
+    @abstractmethod
+    def add_texts_and_embeddings(self, texts: list[Embeddable]) -> None:
+        pass
+
+    @abstractmethod
+    def similarity_search(
+        self, query: list[float], k: int
+    ) -> tuple[list[Embeddable], list[float]]:
+        pass
+
+    @abstractmethod
+    def clear(self) -> None:
+        pass
+
+    def max_marginal_relevance_search(
+        self, query: list[float], k: int, fetch_k: int, lambda_: float = 0.5
+    ) -> tuple[list[Embeddable], list[float]]:
+        """Vectorized implementation of Maximal Marginal Relevance (MMR) search.
+
+        Args:
+            query: Query vector.
+            k: Number of results to return.
+            lambda_: Weighting of relevance and diversity.
```

Comment:
Could you add to the docstring that the lambda can be tuned for different applications of paperqa?

Reply:
done - moved to be part of vectorstore class